io-functions-pay-portal/PostNewslettersRecipients/handler.ts

```diff
@@ -0,0 +1,276 @@
+import { RecipientResponse } from "../generated/definitions/RecipientResponse";
+
+const logPrefix = "PostNewslettersRecipients";
+const config = getConfigOrThrow();
+
+type IPostNewslettersRecipientsHandler = (
+  context: Context,
+  id: NonEmptyString,
+  recipientRequest: RecipientRequest
+) => Promise<IResponseSuccessJson<RecipientResponse> | ErrorResponses>;
+
+/**
+ * Model for mailup OAuth token
+ */
+const AuthTokenR = t.partial({});
+
+const AuthTokenO = t.interface({
+  access_token: NonEmptyString,
+
+  expires_in: t.number,
+
+  refresh_token: t.string
+});
+
+const MailupAuthToken = t.intersection(
+  [AuthTokenR, AuthTokenO],
+  "MailupAuthToken"
+);
+
+type MailupAuthToken = t.TypeOf<typeof MailupAuthToken>;
+
+/**
+ * Model for Google reCaptcha response
+ */
+const ResponseR = t.interface({
+  success: t.boolean,
+
+  challenge_ts: t.string,
+
+  hostname: t.string
+});
+
+const ResponseO = t.partial({
+  "error-codes": t.readonlyArray(t.string, "array of string")
+});
+
+export const ResponseRecaptcha = t.intersection(
+  [ResponseR, ResponseO],
+  "ResponseRecaptcha"
+);
+
+export type ResponseRecaptcha = t.TypeOf<typeof ResponseRecaptcha>;
+
+/**
+ * A middleware that extracts a RecipientRequest payload from a request.
+ */
+const RecipientRequestMiddleware: IRequestMiddleware<
+  "IResponseErrorValidation",
+  RecipientRequest
+> = request =>
+  Promise.resolve(
+    RecipientRequest.decode(request.body).mapLeft(
+      ResponseErrorFromValidationErrors(RecipientRequest)
+    )
+  );
+
+const recaptchaCheckTask = (
+  recaptchaToken: string
+): TaskEither<Error, ResponseRecaptcha> =>
+  tryCatch(
+    () =>
+      fetchApi(`https://www.google.com/recaptcha/api/siteverify`, {
+        body: `secret=${config.RECAPTCHA_SECRET}&response=${recaptchaToken}`,
+        headers: {
+          // tslint:disable-next-line: no-duplicate-string
+          "Content-Type": "application/x-www-form-urlencoded"
+        },
+        method: "POST"
+      }),
+    err => new Error(`Error posting recaptcha API: ${err}`)
+  )
+    .chain(
+      fromPredicate<Error, Response>(
+        r => r.ok,
+        r => new Error(`Error returned from recaptcha API: ${r.status}`)
+      )
+    )
+    .chain(response =>
+      tryCatch(
+        () => response.json(),
+        err => new Error(`Error getting recaptcha API payload: ${err}`)
+      )
+    )
+    .chain(json =>
+      fromEither(
+        ResponseRecaptcha.decode(json).mapLeft(
+          errors =>
+            new Error(
+              `Error while decoding response from recaptcha: ${readableReport(
+                errors
+              )})`
+            )
+        )
+      )
+    )
+    .chain(
+      fromPredicate(
+        ar => ar.success,
+        _ => new Error(`Error checking recaptcha`)
+      )
+    );
+
+const getMailupAuthTokenTask = (): TaskEither<Error, MailupAuthToken> =>
+  tryCatch(
+    () =>
+      fetchApi(`https://services.mailup.com/Authorization/OAuth/Token`, {
```

Comment:
everytime you deal with hardcoded values I suggest to pass it as default values for method params, in this way you can easily override them in unit tests

Reply:
Thanks, i will refactor with the tests 👍